flux-middle/src/expand.rs

```diff
@@ -0,0 +1,387 @@
+//! Expanding the expression alias `defn` in [`flux_middle::fhir`]
```

Comment:
I think `fhir` is not the correct place to do the expansion. I'd rather do it during (right after) [`conv`](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/conv.rs) as a normalization step for types in [`rty`](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/mod.rs#L228). Some reasons why I think this would be better.

1. I imagine this sort of partial evaluation will only become more complicated over time. Complicated type manipulation should go in `rty` to avoid duplication.
2. We already have infrastructure to do [`"folding"`](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/fold.rs) over types in `rty` so we can reuse that. The type [`Binders`](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/mod.rs#L66) can be used to define abstractions, e.g., [here](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/subst.rs#L132) we apply an abstract refinement (of type `Binders<Pred>`).
3. Stuff in `fhir` is meant to capture what the user wrote but in a desugared form. In particular, it was designed to maintain spans. Doing the expansion here breaks this.
4. Doing the normalization eagerly is just one option and it's not clear to me it's the best option. I imagine we can do the normalization lazily in later steps. Having the logic implemented in `rty` keeps our options more open.

More concretely, I think we can `conv` definitions all the way down to `rty`. Then we can `normalize` definitions. We could implement a `normalize` method generically on `TypeFoldable` which I think is really neat.

side note: The function [`replace_bound_vars`](https://github.com/liquid-rust/flux/blob/e03dd11fe7ee909a9cdc7b0c92a06b435eb87026/flux-middle/src/rty/mod.rs#L228) may be more complicated than what we need for this use case because it deals with abstract refinements which are substituted differently. Maybe we can have a dedicated function for when we are only substitution expressions

Reply:
Good point! To make sure I follow, we should 

1. Still do the "wf" checking and such at the `fhir` level (with the rest of the wf), but
2. Propagate the `Defn` down to the `rty` level and then 
3. Do the actual `exp_` at the `rty` level,

yes?

Reply:
Yes, that sounds good.